### PR TITLE
Defining custom decompositions for DQ2

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
The `custom_decomps` and `decomp_depth` arguments when calling `qml.device("default.qubit")` did not work, and as such were not at feature parity with DQL. The `set_decomposition` context also didn't work with the new implementation of `default.qubit`

**Description of the Change:**
Kwargs `decomposer` and `max_expansion` are added to the `preprocessing.decompose` transform. These can be specified to create custom behaviour for decompositions.

A `custom_decomp_preprocessing` function is created, analogous to `create_decomp_expand_fn` for DQL. It defines a custom `decomposer` for `device.decompose` and modifies device `stopping_conditions` based on the `custom_decomps` dictionary provided. It also modifies `device.decompose.max_expansion` if a `decomp_depth` is specified.

There's some special treatment for controlled operators where we override the `op.base.compute_decomposition` method for that instance of the operator if there is a custom decomposition defined for the base.

The tests are all copy-pasted from the old device API tests, with slight modifications where relevant for the new API, and a couple of additional tests that seemed relevant.

**Benefits:**
Feature parity with DQL wrt custom decompositions. It works.

**Possible Drawbacks:**
It feels messy and not very nice (but it does the thing)

### CODE REVIEW

File: pennylane/devices/preprocess.py [hunk 1]
```diff
@@ -213,7 +214,10 @@ def warn_about_trainable_observables(
 def decompose(
     tape: qml.tape.QuantumTape,
     stopping_condition: Callable[[qml.operation.Operator], bool],
-    skip_initial_state_prep: bool = True,
+    skip_initial_state_prep=True,
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:217]:
```suggestion
    skip_initial_state_prep: bool = True,
```

File: pennylane/devices/preprocess.py [hunk 2]
```diff
@@ -224,6 +245,10 @@ def decompose(
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
             a ``DecompositionUndefinedError`` will be raised.
         skip_initial_state_prep=True (bool): If ``True``, the first operator will not be decomposed if it inherits from :class:`~.StatePrepBase`.
+        decomposer (Callable): an optional callable that takes an operator and implements the relevant decomposition.
+            If None, defaults to using a callable returning ``op.decomposition()`` for any :class:`~.Operator` .
+        max_expansion: The maximum depth of the expansion.
```
Comment (trbromley) [inline pennylane/devices/preprocess.py:250]:
```suggestion
        max_expansion (int): The maximum depth of the expansion.
```

File: pennylane/transforms/tape_expand.py [hunk 3]
```diff
@@ -350,6 +350,80 @@ def custom_decomp_expand(self, circuit, max_expansion=decomp_depth):
     return custom_decomp_expand
 
 
+def create_decomp_preprocessing(custom_decomps, dev, decomp_depth=10):
```
Comment (trbromley) [inline pennylane/transforms/tape_expand.py:353]: Is this intended to be user facing?

Comment (lillian542) [reply pennylane/transforms/tape_expand.py:353]: I think no? But is `create_decomp_expand_function` user facing? This is that but for the new device API.

Comment (trbromley) [reply pennylane/transforms/tape_expand.py:353]: `create_decomp_expand_function` is not in the docs :+1:

File: pennylane/devices/preprocess.py [hunk 4]
```diff
@@ -45,21 +46,37 @@ def null_postprocessing(results):
 def _operator_decomposition_gen(
     op: qml.operation.Operator,
     acceptance_function: Callable[[qml.operation.Operator], bool],
+    decomposer: Optional[
+        Callable[[qml.operation.Operator], Sequence[qml.operation.Operator]]
+    ] = None,
+    max_expansion: Optional[int] = None,
+    current_depth=0,
```
Comment (albi3ro) [inline pennylane/devices/preprocess.py:53]: Another totally optional comment.

I wonder if we use the difference between the two and only have one argument.

```
# initialize to np.inf if infinite recursion
depth = np.inf
# check if reached zero yet
depth < 1
# subtract one for next round
depth -= 1
```

File: tests/transforms/test_tape_expand.py [hunk 5]
```diff
@@ -784,7 +841,55 @@ def circuit():
...
+        def circuit():
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(wires=0))
+
+        # Initial test
+        _ = circuit()
+
+        tape = spy.call_args_list[0][0][0][0]
+        assert len(tape.operations) == 1
+        assert tape.operations[0].name == "CNOT"
```
Comment (albi3ro) [inline tests/transforms/test_tape_expand.py:862]: Works fine, but might also be able to use the tracker and resources to get what types of operations got submitted to the circuit. 

```
with dev.tracker:
    _ = circuit()
assert dev.tracker.history['resources'][-1].gate_types['CNOT'] == 1
```

mocker free way of checking this information.

**Review by albi3ro — APPROVED** [review]
Changelog, but other than that! Thanks for putting this together.

File: pennylane/transforms/tape_expand.py (hunk 3 above)
Comment (trbromley) [reply pennylane/transforms/tape_expand.py:353]: Can we have `create_decomp_preprocessing` with a leading underscore?

**Review by mudit2812 — APPROVED** [review]

File: pennylane/transforms/tape_expand.py [hunk 6]
```diff
@@ -350,6 +350,80 @@ def custom_decomp_expand(self, circuit, max_expansion=decomp_depth):
...
+                qml.Hadamard(wires=wires[1])
+            ]
+
+    We then create the custom function (passing a device, in order to pick up any
+    additional stopping criteria the expansion should have), and then register the
+    result as a custom function of the device:
+
+    >>> custom_decomps = {qml.CNOT : custom_cnot}
+    >>> new_preprocessing = create_decomp_preprocessing(custom_decomps, dev)
+    >>> dev.preprocessing = new_preprocessing
```
Comment (timmysilv) [inline pennylane/transforms/tape_expand.py:387]:
```suggestion
    >>> dev.preprocess = new_preprocessing
```

**Review by timmysilv — APPROVED** [review]
does this need a changelog entry? it's adding something that was supposedly (😄) already there

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -214,6 +214,13 @@
   of the computed qubit operator, if imaginary components are smaller than a threshold. 
   [(#4639)](https://github.com/PennyLaneAI/pennylane/pull/4639)
 
+* Updated `qml.device`, `devices.preprocessing` and the `tape_expand.set_decomposition` context 
+  manager to bring `DefaultQubit2` to feature parity with `default.qubit.legacy` with regards to 
+  using custom decompositions. The `DefaultQubit2` device can now be included in a `set_decomposition`
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:227]:
```suggestion
  using custom decompositions. The `DefaultQubit` device can now be included in a `set_decomposition` 
```

Comment (timmysilv) [reply doc/releases/changelog-dev.md:227]: we'll probably need a nice big cleanup for the changelog in this respect, but yep

Comment (lillian542) [reply doc/releases/changelog-dev.md:227]: I thought we were leaving it `DefaultQubit2` everywhere for lack of ambiguity and letting Tom decide a uniform way of describing things when he organizes the changelog?

**Pull request merged**